### PR TITLE
SAAS-2294: Plugin to check AWS Elasticsearch desired instance types

### PR DESCRIPTION
**Pull request by UmairAjmal**

### CODE REVIEW

File: exports.js [hunk 1]
```diff
@@ -195,6 +195,7 @@ module.exports = {
         'esLoggingEnabled'              : require(__dirname + '/plugins/aws/es/esLoggingEnabled.js'),
         'esUpgradeAvailable'            : require(__dirname + '/plugins/aws/es/esUpgradeAvailable.js'),
         'esHttpsOnly'                   : require(__dirname + '/plugins/aws/es/esHttpsOnly.js'),
+        'esDesiredInstanceType'       : require(__dirname + '/plugins/aws/es/esDesiredInstanceType.js'),
```
Comment (AkhtarAmir) [inline exports.js:198]:
```suggestion
        'esDesiredInstanceType'         : require(__dirname + '/plugins/aws/es/esDesiredInstanceType.js'),
```

File: plugins/aws/es/esDesiredInstanceType.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    category: 'ES',
+    description: 'Ensure that all your Amazon Elasticsearch cluster instances are of given instance types.',
+    more_info: 'Limiting the type of Amazon Elasticsearch cluster instances that can be provisioned will help address compliance requirements and prevent unexpected charges on the AWS bill.',
+    link: 'https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-createupdatedomains.html',
+    recommended_action: 'Reconfigure the domain to have the desired instance types.',
+    apis: ['ES:listDomainNames', 'ES:describeElasticsearchDomain', 'STS:getCallerIdentity'],
+    settings: {
+        es_desired_instance_type: {
+            name: 'Desired Instance Type',
+            description: 'Instance type of ES data instances',
```
Comment (AkhtarAmir) [inline plugins/aws/es/esDesiredInstanceType.js:15]: make this data instance

File: plugins/aws/es/esDesiredInstanceType.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                        results, 3,
+                        'Unable to query for ES domain config: ' + helpers.addError(describeElasticsearchDomain), region, resource);
+                    return cb();
+                }
+
+                const localDomainStatus = describeElasticsearchDomain.data.DomainStatus;
+                let disallowedInstanceTypes = [];
+                if (localDomainStatus.ElasticsearchClusterConfig) {
+                    const config = localDomainStatus.ElasticsearchClusterConfig;
+                    if (config.InstanceType && !es_desired_instance_type.includes(config.InstanceType)) disallowedInstanceTypes.push(`${config.InstanceType} type for data node`);
```
Comment (AkhtarAmir) [inline plugins/aws/es/esDesiredInstanceType.js:80]: Move type for data node to result string.

File: plugins/aws/es/esDesiredInstanceType.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                    !describeElasticsearchDomain.data.DomainStatus) {
+                    helpers.addResult(
+                        results, 3,
+                        'Unable to query for ES domain config: ' + helpers.addError(describeElasticsearchDomain), region, resource);
+                    return cb();
+                }
+
+                const localDomainStatus = describeElasticsearchDomain.data.DomainStatus;
+                let disallowedInstanceTypes = [];
+                if (localDomainStatus.ElasticsearchClusterConfig) {
```
Comment (AkhtarAmir) [inline plugins/aws/es/esDesiredInstanceType.js:78]:
```suggestion
                let disallowedInstanceTypes = [];
                if (describeElasticsearchDomain.data.DomainStatus &&
                    describeElasticsearchDomain.data.DomainStatus;.ElasticsearchClusterConfig) {
```

File: plugins/aws/es/esDesiredInstanceType.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+            if (listDomainNames.err || !listDomainNames.data) {
+                helpers.addResult(
+                    results, 3,
+                    'Unable to query for ES domains: ' + helpers.addError(listDomainNames), region);
+                return rcb();
+            }
+
+            if (!listDomainNames.data.length){
+                helpers.addResult(results, 0, 'No ES domains found', region);
```
Comment (AkhtarAmir) [inline plugins/aws/es/esDesiredInstanceType.js:56]: Add !domain.DomainName check

File: plugins/aws/es/esDesiredInstanceType.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    more_info: 'Limiting the type of Amazon Elasticsearch cluster instances that can be provisioned will help address compliance requirements and prevent unexpected charges on the AWS bill.',
+    link: 'https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-createupdatedomains.html',
+    recommended_action: 'Reconfigure the domain to have the desired instance types.',
+    apis: ['ES:listDomainNames', 'ES:describeElasticsearchDomain', 'STS:getCallerIdentity'],
+    settings: {
+        es_desired_instance_type: {
+            name: 'Desired Instance Type',
+            description: 'Instance type of ES data instances',
+            regex: '^.*$',
+            default: 't2.small.elasticsearch'
```
Comment (AkhtarAmir) [inline plugins/aws/es/esDesiredInstanceType.js:17]: Remove default values

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/es/esDesiredInstanceType.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    apis: ['ES:listDomainNames', 'ES:describeElasticsearchDomain', 'STS:getCallerIdentity'],
+    settings: {
+        es_desired_data_instance_type: {
+            name: 'ES Data Instance Type',
+            description: 'Instance type of ES data instances',
+            regex: '^.*$',
+            default: ''
+        },
+        es_desired_master_instance_type: {
+            name: 'ES Dedicated Master Instance Type',
```
Comment (AkhtarAmir) [inline plugins/aws/es/esDesiredInstanceType.js:20]:
```suggestion
            name: 'ES Master Instance Type',
```

File: plugins/aws/es/esDesiredInstanceType.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+                return rcb();
+            }
+
+            if (!listDomainNames.data.length){
+                helpers.addResult(results, 0, 'No ES domains found', region);
+                return rcb();
+            }
+
+            for (const domain of listDomainNames.data) {
+                if (!domain.DomainName) return;
```
Comment (AkhtarAmir) [inline plugins/aws/es/esDesiredInstanceType.js:61]:
```suggestion
                if (!domain.DomainName) continue;
```

File: plugins/aws/es/esDesiredInstanceType.js [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+                const resource = `arn:${awsOrGov}:es:${region}:${accountId}:domain/${domain.DomainName}`;
+
+                if (!describeElasticsearchDomain ||
+                    describeElasticsearchDomain.err ||
+                    !describeElasticsearchDomain.data ||
+                    !describeElasticsearchDomain.data.DomainStatus) {
+                    helpers.addResult(
+                        results, 3,
+                        'Unable to query for ES domain config: ' + helpers.addError(describeElasticsearchDomain), region, resource);
+                    return;
```
Comment (AkhtarAmir) [inline plugins/aws/es/esDesiredInstanceType.js:75]:
```suggestion
                    continue;
```

File: plugins/aws/es/esDesiredInstanceType.js [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+                const masterStatus = (disallowedDedicatedInstanceTypes.length) ? 2 : 0;
+                helpers.addResult(results, masterStatus,
+                    `ES cluster is using ${masterStatus == 0 ? 'allowed' : disallowedDedicatedInstanceTypes.join(', ')} master instance types`,
+                    region, resource);
+            }
+
+            rcb();
+
+        }, function() {
```
Comment (AkhtarAmir) [inline plugins/aws/es/esDesiredInstanceType.js:99]:
```suggestion
            rcb();
        }, function() {
```

**Review by AkhtarAmir — APPROVED** [review]

**Review by giorod3 — APPROVED** [review]

**Pull request merged**